### PR TITLE
[sitecore-jss] Optimize initial Retry delay time

### PR DESCRIPTION
**Pull request by sc-addypathania**
At present, the initial retry delay evaluated to 2 seconds, which becomes relatively high for subsequent retries when assessed in an exponential manner. We aim to initiate the first retry after a 1-second wait instead.

- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated

## Testing Details

- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
I assume that it should affect some unit tests, did you check? is it covered?

File: packages/sitecore-jss/src/graphql-request-client.ts [hunk 1]
```diff
@@ -113,7 +113,7 @@ export class DefaultRetryStrategy implements RetryStrategy {
     const rawHeaders = error.response?.headers;
     const delaySeconds = rawHeaders?.get('Retry-After')
       ? Number.parseInt(rawHeaders?.get('Retry-After'), 10)
-      : Math.pow(this.factor, attempt);
+      : Math.pow(this.factor, attempt - 1);
```
Comment (art-alexeyenko) [inline packages/sitecore-jss/src/graphql-request-client.ts:116]: do we have any tests to ensure delay is correctly calculated? similar to "should use [retry-after] header value when response is 429" case
I think that's what Illia meant by additional tests

**Review by art-alexeyenko — CHANGES_REQUESTED** [review]

File: packages/sitecore-jss/src/graphql-request-client.ts (hunk 1 above)
Comment (sc-addypathania) [reply packages/sitecore-jss/src/graphql-request-client.ts:116]: the last test i added above tests the function getDelay if it return the value correctly.

**Review by art-alexeyenko — APPROVED** [review]

File: packages/sitecore-jss/src/graphql-request-client.test.ts [hunk 2]
```diff
@@ -405,4 +406,40 @@ describe('GraphQLRequestClient', () => {
...
+      const shouldRetry = retryStrategy.shouldRetry(mockClientError, 2, 1);
+      expect(shouldRetry).to.equal(false);
+    });
+
+    it('should return delay using exponential backoff when Retry-After header is not present', () => {
+      const delay = retryStrategy.getDelay(mockClientError, 3);
+      const expectedDelay = Math.pow(retryStrategy['factor'], 3 - 1) * 1000;
+      expect(delay).to.equal(expectedDelay);
+    });
+  });
```
Comment (ambrauer) [inline packages/sitecore-jss/src/graphql-request-client.test.ts:444]: There are a few more tests we could add here for better coverage, for example:
- shouldRetry returns false when retries is 0
- shouldRetry returns false when attempt exceeds retries
- shouldRetry returns false when error response status undefined
- getDelay uses custom exponential factor

**Review by ambrauer — APPROVED** [review]

File: packages/sitecore-jss/src/graphql-request-client.ts [hunk 3]
```diff
@@ -92,12 +92,13 @@ export class DefaultRetryStrategy implements RetryStrategy {
...
-   * @param {number[]} statusCodes HTTP status codes to trigger retries on
-   * @param {number} factor Factor by which the delay increases with each retry attempt
+   * @param {Object} options Configurable options for retry mechanism.
+   * @param {number[]} options.statusCodes HTTP status codes to trigger retries on
+   * @param {number} options.factor Factor by which the delay increases with each retry attempt
    */
-  constructor(statusCodes?: number[], factor?: number) {
-    this.statusCodes = statusCodes || [429];
-    this.factor = factor || 2;
+  constructor(options: { statusCodes?: number[]; factor?: number }) {
```
Comment (ambrauer) [inline packages/sitecore-jss/src/graphql-request-client.ts:99]: Need a default empty object to make `options` optional. Probably good to add a `DefaultRetryStrategy` unit test to cover that. e.g. 'should use statusCode 429 and factor 2 by default'
```suggestion
  constructor(options: { statusCodes?: number[]; factor?: number } = {}) {
```

**Pull request merged**